src/protocols/cosmic_workspaces/cosmic-workspaces.c

```diff
@@ -0,0 +1,675 @@
+// SPDX-License-Identifier: GPL-2.0-only
+#include <assert.h>
+#include <wlr/types/wlr_output.h>
+#include <wlr/util/log.h>
+#include "common/mem.h"
+#include "common/list.h"
+#include "cosmic-workspace-unstable-v1-protocol.h"
+#include "protocols/cosmic-workspaces.h"
+#include "protocols/cosmic-workspaces-internal.h"
+
+/*
+ *	.--------------------.
+ *	|        TODO        |
+ *	|--------------------|
+ *	| - destroy all ws   |
+ *	|   on group destroy |
+ *	| - convert done evt |
+ *	|   to switch()      |
+ *	| - prevent empty    |
+ *	|   done events      |
+ *	| - go through xml   |
+ *	|   and verify impl  |
+ *	| - assert pub API   |
+ *	| - maybe split into |
+ *	|   more files       |
+ *	`--------------------´
+ *
+ */
+
+#define COSMIC_WORKSPACE_V1_VERSION 1
+
+/* These are just *waaay* too long */
+#define ZCOSMIC_CAP_WS_CREATE \
+	ZCOSMIC_WORKSPACE_GROUP_HANDLE_V1_ZCOSMIC_WORKSPACE_GROUP_CAPABILITIES_V1_CREATE_WORKSPACE
+#define ZCOSMIC_CAP_WS_ACTIVATE \
+	ZCOSMIC_WORKSPACE_HANDLE_V1_ZCOSMIC_WORKSPACE_CAPABILITIES_V1_ACTIVATE
+#define ZCOSMIC_CAP_WS_DEACTIVATE \
+	ZCOSMIC_WORKSPACE_HANDLE_V1_ZCOSMIC_WORKSPACE_CAPABILITIES_V1_DEACTIVATE
+#define ZCOSMIC_CAP_WS_REMOVE \
+	ZCOSMIC_WORKSPACE_HANDLE_V1_ZCOSMIC_WORKSPACE_CAPABILITIES_V1_REMOVE
+
+enum workspace_state {
+	CW_WS_STATE_ACTIVE  = 1 << 0,
+	CW_WS_STATE_URGENT  = 1 << 1,
+	CW_WS_STATE_HIDDEN  = 1 << 2,
+
+	/*
+	 * Set when creating a new workspace so we
+	 * don't end up having to send the state twice.
+	 */
+	CW_WS_STATE_INVALID = 1 << 31,
+};
+
+#define wl_arr_add(_arr, _val) (*(__typeof__(_val) *)wl_array_add( \
```

Comment:
It's not obvious to me what `wl_arr_add()` does. Should we add to `include/common/array.h` and add a comment to describe it.

Reply:
Created a new helper in `common/array.h`: `xwl_array_add()` (in the style of `xstrdup()`) which asserts against memory allocation errors.